source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOVisibilityManager.mm

```diff
@@ -7,112 +7,212 @@
+/// adds index of a visible view to a visible views collection, and the index is maintained in sorted order
+/// in increasing order.
+- (void)addVisibleView:(NSUInteger)index
 {
-    if (!subviews || !viewToBeHidden) {
+    NSNumber *indexAsNumber = [NSNumber numberWithLong:index];
+    if ([_visibleViews containsObject:indexAsNumber]) {
         return;
     }
+    NSRange range = NSMakeRange(0, _visibleViews.count);
+    NSUInteger insertionIndex = [_visibleViews indexOfObject:indexAsNumber
+                                               inSortedRange:range
+                                                     options:NSBinarySearchingInsertionIndex
+                                             usingComparator:^(id num0, id num1) {
+                                                 return [num0 compare:num1];
+                                             }];
+    [_visibleViews insertObject:indexAsNumber atIndex:insertionIndex];
+}
 
-    if (!viewToBeHidden.isHidden) {
-        viewToBeHidden.hidden = YES;
+/// removes the index of view from the visible views collection
+/// maintain the sorted order after the removal
+- (void)removeVisibleViewAtIndex:(NSUInteger)index
+{
+    NSNumber *indexAsNumber = [NSNumber numberWithLong:index];
+    if (![_visibleViews containsObject:indexAsNumber]) {
+        return;
     }
+    NSRange range = NSMakeRange(0, _visibleViews.count);
+    NSUInteger removalIndex = [_visibleViews indexOfObject:indexAsNumber
+                                             inSortedRange:range
+                                                   options:NSBinarySearchingInsertionIndex
+                                           usingComparator:^(id num0, id num1) {
+                                               return [num0 compare:num1];
+                                           }];
+    [_visibleViews removeObjectAtIndex:removalIndex];
+}
 
-    // get the index of the target view
-    // if the target view is not the tail,
-    // check to the right, and if the right separator and not hidden,
-    // hide the separator
-    // if the target view is tail, check to left, and if the left is separator and not hidden,
-    // hide the separator
-    NSInteger index = [subviews indexOfObject:viewToBeHidden];
-    if (index + 1 < subviews.count) {
-        NSInteger nextViewIndex = index + 1;
-        UIView *nextView = subviews[nextViewIndex];
-        if ([nextView isKindOfClass:[ACRSeparator class]]) {
-            if (!nextView.isHidden) {
-                nextView.hidden = YES;
-            }
-        }
-    } else if (index - 1 >= 0) {
-        NSInteger prevViewIndex = index - 1;
-        UIView *prevView = subviews[prevViewIndex];
-        if ([prevView isKindOfClass:[ACRSeparator class]]) {
-            if (!prevView.isHidden) {
-                prevView.hidden = YES;
-            }
-        }
+/// YES means the index of view is currently the leading or top view
+- (BOOL)amIHead:(NSUInteger)index
```

Comment:
>amIHead [](http://example.com/codeflow?start=8&length=7)

possible rename to isIndexedItemHead/Leading/Top or something similar

Reply:
thanks for the suggestion.  i will keep it as it is as the suggested change doesn't drastically improve what's here.